src/message_cache.rs

```diff
@@ -0,0 +1,70 @@
+use crate::Result;
+use std::{
+    collections::VecDeque,
+    ops::Deref,
+    time::{Duration, Instant},
+};
+
+pub struct MessageCache<T> {
+    waiting: VecDeque<CacheMessage<T>>,
+    max_messages: u16,
+}
+
+#[derive(Debug)]
+pub struct CacheMessage<T> {
+    received: Instant,
+    message: T,
+}
+
+impl<T> CacheMessage<T> {
+    pub fn hold_time(&self) -> Duration {
+        self.received.elapsed()
+    }
+
+    pub fn into_inner(self) -> T {
+        self.message
+    }
+}
+
+impl<T> Deref for CacheMessage<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.message
+    }
+}
+
+impl<T> MessageCache<T> {
+    pub fn new(max_messages: u16) -> Self {
+        let waiting = VecDeque::new();
+        Self {
+            waiting,
+            max_messages,
+        }
+    }
+
+    pub fn store(&mut self, message: T, received: Instant) -> Result {
+        self.waiting.push_back(CacheMessage { message, received });
+        if self.waiting_count() > self.max_messages as usize {
+            self.waiting.pop_front();
+        }
```

Comment:
This assumes GC has ran since the last time this function was called. It's probably better to either:

1. inline the GC here and have this fn return number of GCs messages
2. not do any popping here if the number of GCd messages is actually important to know

Reply:
Fixed in [433ce5b](https://github.com/helium/gateway-rs/pull/364/commits/433ce5bf279a402bae426365020d265f03aca024)